solver/src/naive_amm_settlement.rs

```diff
@@ -0,0 +1,301 @@
+        .collect(),
+        trades: vec![
+            Trade {
+                order: *sell_a,
+                executed_amount: sell_a.sell_amount,
+            },
+            Trade {
+                order: *sell_b,
+                executed_amount: sell_b.sell_amount,
+            },
+        ],
+        ..Default::default()
+    }
+}
+
+// Match two orders with amm assuming that there is price overlap.
+fn amm_match_(sell_a: &UserOrder, sell_b: &UserOrder) -> Option<Settlement> {
+    // Based on our assumptions we know that exactly one order is "bigger" than the other in the
+    // sense that it a larger sell amount than the other order's buy amount.
+    // It is not possible for both orders to be bigger because that would be a direct match which
+    // has already been handled.
+    let (big, small) = if sell_a.sell_amount > sell_b.buy_amount {
+        (sell_a, sell_b)
+    } else {
+        (sell_b, sell_a)
+    };
+    // Unwrap because of the above explanation.
+    let bigger_sing_buy_amount = big.buy_amount.checked_sub(small.sell_amount).unwrap();
+    // Because the smart contract enforces uniform prices we must pick one price that is accepted by
+    // both orders. We know that there is price overlap so we could pick any price between the price
+    // limits of the orders. We pick the price of the bigger order because it's sell token surplus
+    // will have to be traded with the amm and the bigger order always has a more lax amm price
+    // requirement of the bigger order's sell token than the smaller order.
+    // Because we picked the bigger order's price we must add an extra buy amount to the smaller
+    // order so that it gets the same price.
+    let small_buy_amount = rounded_up_division(
+        small.sell_amount.checked_mul(big.sell_amount)?,
+        big.buy_amount,
+    )?;
+    // Does this have to be saturating or is guaranteed to not underflow? With the rounding from the
+    // previous operation I suspect it might.
+    let big_extra_sell_amount = big.sell_amount.saturating_sub(small_buy_amount);
+    if big_extra_sell_amount.is_zero() {
+        return None;
+    }
+
+    Some(Settlement {
+        clearing_prices: [
+            (big.sell_token, big.sell_amount),
+            (big.buy_token, big.buy_amount),
+        ]
+        .iter()
+        .copied()
+        .collect(),
+        trades: vec![
+            Trade {
+                order: *sell_a,
+                // TODO: or do we set this to 0 to save a tiny bit of gas because smart contract
+                // assumes correct sell amount anyway for fill-or-kill order.
+                executed_amount: sell_a.sell_amount,
+            },
+            Trade {
+                order: *sell_b,
+                executed_amount: sell_b.sell_amount,
+            },
+        ],
+        interactions: vec![Interaction::AmmSwapExactTokensForTokens {
+            amount_in: big_extra_sell_amount,
+            amount_out_min: bigger_sing_buy_amount,
```

Comment:
If I understand you right, you're saying this should be 0 because we might be able to get the required amount from the buffer in the contract. I do think as you say at the end this should be handled by another layer that has access to the current state of the block chain for the amm pool and the state of the buffer.